### PR TITLE
More Kinesis resharding adjustments

### PR DESCRIPTION
**Pull request by jon-wei**
This PR adjusts certain aspects of the Kinesis resharding fixes introduced in #8644:
- #8644 changed the shardID -> partition group mapping to use hashing, to provide a fixed mapping. However, when shards can expire, this can result in imbalanced distribution of active shards among tasks (e.g., some tasks may not get any readable shards). This PR reverts the mapping logic back to using the position of the shard within the shard list % task count, restricting the shard list to only non-expired shards. To handle expired shards, this PR instead recomputes the partition group mappings when shards expire, and tells any active tasks created under the old mappings to exit early.
- Instead of deleting expired shards from metadata, this PR changes the expiration handling to store an "EXPIRED" sequence number marker instead. While expired shards should never reappear, keeping a record of expired shards in metadata will allow us to better handle such anomalies.
- Adjusts `checkSequenceNumberAvailability` (renamed to `checkOffsetAvailability`), only the earliest available offset needs to be compared against the target offset. Because the "LATEST" Kinesis shard iterator only provides access to the seq no after the most recent record (and there is no mode to get the latest readable record), the "get latest offset" operation in Kinesis runs in a polling loop. When data is not being continuously written to a shard, this can result in a Kinesis supervisor making extremely slow progress as it waits for the polling loop to timeout for each shard that has no recent data. Removing the latest offset check restricts this problem situation to only when a Kinesis supervisor first discovers a shard that's empty (after data is written to that shard once, the earliest seq no check can return immediately)

This PR has:
- [x] been self-reviewed.
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java [hunk 1]
```diff
@@ -1946,53 +1956,152 @@ private boolean updatePartitionDataFromStream()
...
-        if (!partitionIds.contains(partitionTd)) {
-          expiredPartitions.add(partitionTd);
+    if (!partitionIds.equals(previousPartitionIds)) {
+      // the set of partition IDs has changed, have any running tasks stop early so that we can adjust to the
+      // repartitioning quickly by creating new tasks
+      for (TaskGroup taskGroup : activelyReadingTaskGroups.values()) {
+        if (!taskGroup.taskIds().isEmpty()) {
+          // a new partition was added and we are managing active tasks - set an early publish time 2 minutes in the
+          // future to give things time to settle
+          earlyStopTime = DateTimes.nowUtc().plusMinutes(2);
```
Comment (clintropolis) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java]: Where did this number come from, should it be a configuration, or is there any thing else like this in other parts of the codebase?

File: extensions-core/kinesis-indexing-service/src/test/java/org/apache/druid/indexing/kinesis/supervisor/KinesisSupervisorTest.java [hunk 2]
```diff
@@ -3885,8 +3906,8 @@ public void testShardSplit() throws Exception
         SHARD_ID0,
         "0"
     ));
-    // there would be 4 tasks, 2 for each task group
-    EasyMock.expect(taskClient.getCheckpointsAsync(EasyMock.contains("sequenceName-1"), EasyMock.anyBoolean()))
+    // there would be 1 task, since there is only 1 shard
```
Comment (clintropolis) [inline extensions-core/kinesis-indexing-service/src/test/java/org/apache/druid/indexing/kinesis/supervisor/KinesisSupervisorTest.java]: :+1:

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java (hunk 1 above)
Comment (jon-wei) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java]: I updated this to use a configurable `repartitionTransitionDuration` property under the supervisor tuning config, along with docs on what the property is for (giving the stream time to write records to the new shards before transitioning to a new set of tasks)

Comment (jon-wei) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java]: It makes sense to support that for Kafka as well (partitions can be added), but I'll do that in a follow-on PR

**Pull request merged**